modules/go/main.go

```diff
@@ -34,57 +110,280 @@ type Go struct {
+//   - Apply configuration
 //   - Mount the source code
-//   - Download dependencies
-func (p *Go) Env(
-	// Any extra apk packages that should be included in the base image
+func (p Go) Env(
 	// +optional
-	extraPackages []string,
+	platform dagger.Platform,
 ) *dagger.Container {
-	return p.
-		Base(extraPackages).
-		WithEnvVariable("CGO_ENABLED", "0").
-		WithWorkdir("/app").
-		// run `go mod download` with only go.mod files (re-run only if mod files have changed)
-		WithDirectory("/app", p.Source, dagger.ContainerWithDirectoryOpts{
-			Include: []string{"**/go.mod", "**/go.sum"},
+	return p.Base.
+		// Configure CGO
+		WithEnvVariable("CGO_ENABLED", func() string {
+			if p.Cgo {
+				return "1"
+			}
+			return "0"
+		}()).
+		// Configure platform
+		With(func(c *dagger.Container) *dagger.Container {
+			if platform == "" {
+				return c
+			}
+			spec := platforms.Normalize(platforms.MustParse(string(platform)))
+			c = c.
+				WithEnvVariable("GOOS", spec.OS).
+				WithEnvVariable("GOARCH", spec.Architecture)
+			switch spec.Architecture {
+			case "arm", "arm64":
+				switch spec.Variant {
+				case "", "v8":
+				default:
+					c = c.WithEnvVariable("GOARM", strings.TrimPrefix(spec.Variant, "v"))
+				}
+			}
+			return c
 		}).
-		WithExec([]string{"go", "mod", "download"}).
-		// run `go build` with all source
-		WithMountedDirectory("/app", p.Source)
+		WithMountedDirectory("", p.Source)
+}
+
+// List tests
+func (p Go) Tests(
+	ctx context.Context,
+	// Packages to list tests from (default all packages)
+	// +optional
+	// +default=["./..."]
+	pkgs []string,
+) (string, error) {
+	script := "go test -list=. " + strings.Join(pkgs, " ") + " | grep ^Test | sort"
+	return p.
+		Env(defaultPlatform).
+		WithExec([]string{"sh", "-c", script}).
+		Stdout(ctx)
+}
+
+// Build the given main packages, and return the build directory
+func (p Go) Build(
+	ctx context.Context,
+	// Which targets to build (default all main packages)
+	// +optional
+	// +default=["./..."]
+	pkgs []string,
+	// Disable symbol table
+	// +optional
+	noSymbols bool,
+	// Disable DWARF generation
+	// +optional
+	noDwarf bool,
+	// Target build platform
+	// +optional
+	platform dagger.Platform,
+	// Output directory
+	// +optional
+	// +default="./bin/"
+	output string,
+) (*dagger.Directory, error) {
+	if p.Race {
+		p.Cgo = true
+	}
+	mainPkgs, err := p.ListPackages(ctx, pkgs, true)
+	if err != nil {
+		return nil, err
+	}
+	ldflags := p.Ldflags
+	if noSymbols {
+		ldflags = append(ldflags, "-s")
+	}
+	if noDwarf {
+		ldflags = append(ldflags, "-w")
+	}
+	env := p.
+		Download().
+		Env(platform)
+	cmd := []string{"go", "build", "-o", output}
+	for _, pkg := range mainPkgs {
+		env = env.WithExec(goCommand(cmd, []string{pkg}, ldflags, p.Values, p.Race))
+	}
+	return dag.Directory().WithDirectory(output, env.Directory(output)), nil
+}
+
+// Build a single main package, and return the compiled binary
+func (p Go) Binary(
```

Comment:
While we're here - what do you think about `Executable` instead of `Binary`? `Binary` feels too generic still (it's better than `File` which is what it did used to be ages ago).

Reply:
In the context of CI/CD, I don't think "binary" is too generic. Here's what I get from gpt4:

> > In a devops and CI/CD context, what is a "binary"?
> In DevOps and CI/CD, a “binary” refers to an executable file produced by compiling source code. This binary file contains machine code that a computer can run directly without needing an interpreter or further compilation. In a CI/CD pipeline, binaries are often built as artifacts during the build stage and then tested, deployed, or distributed as part of software delivery.

Reply:
Sure! No objections, was just wondering if we wanted to change this while in the area - sounds good as is :tada: